File: CA4/Q1.cpp
#include <iostream>
#include <tuple>
#include <vector>
#include <queue>
using namespace std;


struct Node 
{
	tuple <int, int, int> data;
	Node* next;
	Node(tuple <int, int, int> d)
	{
		data = d;
		next = NULL;
	}
};

struct Queue {
	Node *front, *rear;
	Queue()
	{
		front = rear = NULL;
	}
    //x arguments: (x,y,time)
	void enQueue(tuple <int, int, int> x)
	{
		Node* temp = new Node(x);

		if (rear == NULL) {
			front = rear = temp;
			return;
		}

		rear->next = temp;
		rear = temp;
	}

	void deQueue()
	{
		if (front == NULL)
			return;

		Node* temp = front;
		front = front->next;
    
		if (front == NULL)
			rear = NULL;

		delete (temp);
	}
};

class PristonBuilding
{
public:
    PristonBuilding(int _m,int _n,int _k) : m(_m), n(_n), k(_k){};
    void new_building(tuple <int, int, int> d)
    {
        q.push(d);
    }

    int start_simulation()
    {
        vector<vector<int>> matrix(n, std::vector<int>(m, 0));
        int directions[4][2] = {{-1,0},{1,0},{0,-1},{0,1}};

        int time = 1;

        while(!q.empty())
        {
            tuple <int, int, int> current_node = q.front();
            q.pop();

            matrix[get<0>(current_node)][get<1>(current_node)] = 1;
            for(int i = 0;i < 4;i++)
            {
                tuple <int, int,int> adj_node;
                adj_node = make_tuple(directions[i][0] + get<0>(current_node), directions[i][1] + get<1>(current_node),0);
                if(!is_valid(adj_node))
                    continue;
                 if(matrix[get<0>(adj_node)][get<1>(adj_node)] == 1)
                    continue;
                time = get<2>(current_node) + 1;
                get<2>(adj_node) = time;
                matrix[get<0>(adj_node)][get<1>(adj_node)] = 1;
                new_building(adj_node);
            }
        }

        return time;
    }
private:
    int m;
    int n;
    int k;
    queue<tuple <int, int, int>> q;

    bool is_valid(tuple <int, int, int> pos)
    {
        return get<0>(pos) < n && get<1>(pos) < m and get<0>(pos) >= 0 && get<1>(pos) >= 0;
    }

};

void run()
{
    int m,n,k;
    cin >> m >> n >> k;
    PristonBuilding pb(m,n,k);

    int i = 0;
    while(i < 2*k)
    {
        int x,y;
        cin >> x >> y;
        tuple <int, int, int> building;
        building = make_tuple(x-1,y-1,1);
        pb.new_building(building);
        i += 2;
    }

    cout << pb.start_simulation() << endl;
}
int main()
{
    run();

}

File: CA4/Q1.py
from collections import deque 
from pprint import pprint

class PristonBuilding :
    def __init__(self,m,n,k) :
        self.m = m
        self.n = n
        self.k = k
        self.q = deque()
        self.matrix = [[0]*m for _ in range(n)]
    
    def new_building(self,x,y,time) :
        self.q.append((x,y,time))
        
    def is_valid(self,pos) :
        return pos[0] < self.n and pos[1] < self.m and pos[0] >= 0 and pos[1] >= 0

    def start_simulation(self) :
        directions = [(-1,0),(1,0),(0,-1),(0,1)]
        time = 1
        while self.q :
            current_node = self.q.popleft()
    
            self.matrix[current_node[0]][current_node[1]] = 1
            for direction in directions :
                adj_node = (direction[0] + current_node[0],direction[1] + current_node[1])
                if not self.is_valid(adj_node) : continue
                if self.matrix[adj_node[0]][adj_node[1]] == 1 : continue
                
                time = current_node[2] + 1
                
                self.matrix[adj_node[0]][adj_node[1]] = 1  #Burnt
                
                self.new_building(adj_node[0],adj_node[1],time)
                
#                 pprint(self.matrix)
#                 print(time)
#                 print("\n")
#                 print(adj_node)
        
        return time
                


# In[96]:


# m = 10
# n = 10
# k = 4
# pb = PristonBuilding(m,n,k)

# #5 3 4 7 7 5 8 5
# pb.new_building(5-1,3-1,1)
# pb.new_building(4-1,7-1,1)
# pb.new_building(7-1,5-1,1)
# pb.new_building(8-1,5-1,1)

# # pb.new_building(0,0,1)

# print(pb.start_simulation())

# print(pb.matrix)


def cin() :
    string = input().strip()
    while string == "" :
        string = input().strip()
    return string

def run() :
    n,m = cin().split()
    k = int(cin())
    pb = PristonBuilding(int(m),int(n),k)
    
    cells = cin().split()

    i = 0
    while i < 2*k :
        pb.new_building(int(cells[i]) - 1,int(cells[i+1]) - 1,1)
        i += 2
    
    print(pb.start_simulation())
        

run()